File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin() , nums.end());     
        if(nums.size() < 3){                
            return {};
        }
        if(nums[0] > 0){                    
            return {};
        }
        vector<vector<int>> ans;
        unordered_map<int,int> hashMap;
        for(int i = 0 ; i < nums.size() ; ++i){     
            hashMap[nums[i]] = i;
        }
        for(int i = 0 ; i < nums.size() - 2 ; ++i){     
            if(nums[i] > 0) {     
                break;
            }
            for(int j = i + 1 ; j < nums.size() - 1 ; ++j) {
                int required = -1*(nums[i] + nums[j]);   
                if(hashMap.count(required) && hashMap.find(required)->second > j) { 
                    ans.push_back({nums[i] , nums[j] , required});
                }
                j = hashMap.find(nums[j])->second;
            }
            i = hashMap.find(nums[i])->second;
        }
        return ans;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        for i in range(len(nums)):
            if nums[i] > 0:
                break
            if i > 0 and nums[i] == nums[i-1]:
                continue

            j = i + 1
            k = len(nums) - 1
            while j < k:
                sum_ = nums[i] + nums[j] + nums[k]
                if sum_ == 0:
                    res.append([nums[i],nums[j],nums[k]])
                    j += 1
                    k -= 1

                    while j < k and nums[j] == nums[j-1]:
                        j += 1
                    while j < k and nums[k] == nums[k+1]:
                        k -= 1
                else:
                    if sum_ > 0:
                        k -= 1
                    if sum_ < 0:
                        j += 1
        
        return res
                
        